lib/protoflow-blocks/src/blocks/text/encode_csv.rs

```diff
@@ -0,0 +1,135 @@
+// This is free and unencumbered software released into the public domain.
+
+extern crate std;
+
+use crate::{
+    prelude::{Bytes, String, ToString, Vec},
+    StdioConfig, StdioError, StdioSystem, System,
+};
+use protoflow_core::{
+    types::{Value, value::Kind::*}, Block, BlockError, BlockResult, BlockRuntime, InputPort, OutputPort,
+};
+use protoflow_derive::Block;
+use simple_mermaid::mermaid;
+use csv::WriterBuilder;
+use std::io::Cursor;
+/// A block that encodes CSV files by converting a header and rows, provided as `prost_types::Value` streams, into a byte stream.
+///
+/// # Block Diagram
+#[doc = mermaid!("../../../doc/text/encode_csv.mmd")]
+///
+/// # Sequence Diagram
+#[doc = mermaid!("../../../doc/text/encode_csv.seq.mmd" framed)]
+///
+/// # Examples
+///
+/// ## Using the block in a system
+///
+/// ```rust
+/// # use protoflow_blocks::*;
+/// # fn main() {
+/// System::build(|s| {
+///     // TODO
+/// });
+/// # }
+/// ```
+///
+/// ## Running the block via the CLI
+///
+/// ```console
+/// $ protoflow execute EncodeCsv
+/// ```
+///
+#[derive(Block, Clone)]
+pub struct EncodeCsv {
+    /// The header message proto_types::Value stream.
+    #[input]
+    pub header: InputPort<Value>,
+    /// The rows message proto_types::Value stream.
+    #[input]
+    pub rows: InputPort<Value>,
+    /// The output message stream.
+    #[output]
+    pub output: OutputPort<Bytes>,
+}
+
+impl EncodeCsv {
+    pub fn new(header: InputPort<Value>, rows: InputPort<Value>, output: OutputPort<Bytes>) -> Self {
+        Self { header, rows, output }
+    }
+
+    pub fn with_system(system: &System) -> Self {
+        use crate::SystemBuilding;
+        Self::new(system.input(), system.input(), system.output())
+    }
+}
+
+impl Block for EncodeCsv {
+    fn execute(&mut self, runtime: &dyn BlockRuntime) -> BlockResult {
+        runtime.wait_for(&self.header)?;
+        runtime.wait_for(&self.rows)?;
```

Comment:
Don't wait for both ports before proceeding, or you increase the possibility of deadlocks. Just receive the header message, and after that enter the rows loop. No waits needed, really.